### PR TITLE
Register default shader module with Deck

### PR DESCRIPTION
**Pull request by Pessimistress**
For #8680

#### Change List
- Replace `ShaderAssembler.addDefaultModule` with a method on Deck class
- Fix shadow module minor issues
- Render tests

### CODE REVIEW

File: modules/core/src/lib/layer-manager.ts [hunk 1]
```diff
@@ -75,6 +74,8 @@ export default class LayerManager {
   private _needsUpdate: string | false = false;
   private _nextLayers: LayersList | null = null;
   private _debug: boolean = false;
+  // This flag is sepaprate from _needsUpdate because it can be set during an update and should trigger another full update
```
Comment (felixpalmer) [inline modules/core/src/lib/layer-manager.ts:77]:
```suggestion
  // This flag is separate from _needsUpdate because it can be set during an update and should trigger another full update
```

File: modules/core/src/shaderlib/shadow/shadow.ts [hunk 2]
```diff
@@ -190,7 +190,9 @@ function createShadowUniforms(
   if (!shadowEnabled || !opts.shadowMatrices || !opts.shadowMatrices.length) {
     return {
       shadow_uDrawShadowMap: false,
-      shadow_uUseShadowMap: false
+      shadow_uUseShadowMap: false,
+      shadow_uShadowMap0: opts.dummyShadowMap,
```
Comment (felixpalmer) [inline modules/core/src/shaderlib/shadow/shadow.ts:194]: Why is this required?

File: test/modules/core/effects/lighting-effect.spec.ts [hunk 3]
```diff
@@ -13,6 +13,16 @@ const testViewport = new MapView().makeViewport({
   viewState: {longitude: -122, latitude: 37, zoom: 13}
 });
 
+function makeMockContext(device, layerManager) {
+  return {
+    device,
+    deck: {
+      _addDefaultShaderModule: layerManager.addDefaultShaderModule.bind(layerManager),
```
Comment (felixpalmer) [inline test/modules/core/effects/lighting-effect.spec.ts:20]: Could avoid this by doing: https://github.com/visgl/deck.gl/pull/8681#discussion_r1531718265

File: test/modules/extensions/terrain/terrain-cover.spec.ts [hunk 4]
```diff
@@ -147,7 +147,7 @@ test('TerrainCover#viewport diffing#non-geo', async t => {
   t.end();
 });
 
-test('TerrainCover#layers diffing#non-geo', async t => {
+test.skip('TerrainCover#layers diffing#non-geo', async t => {
```
Comment (felixpalmer) [inline test/modules/extensions/terrain/terrain-cover.spec.ts:150]: Just double-checking, are these skipped intentionally?

**Review by felixpalmer — APPROVED** [review]

File: modules/core/src/shaderlib/shadow/shadow.ts (hunk 2 above)
Comment (Pessimistress) [reply modules/core/src/shaderlib/shadow/shadow.ts:194]: v9 throws if a sampler is left unbound, even if it is not used.

File: test/modules/extensions/terrain/terrain-cover.spec.ts (hunk 4 above)
Comment (Pessimistress) [reply test/modules/extensions/terrain/terrain-cover.spec.ts:150]: Yes, the current harness needs some updates to get default effect registration working. I will make the change in a follow-up PR.
